First version

jhford opened the pull request:
This is a library which does verification of Amazon EC2 Instance Identity Documents.  We're taking this approach instead of using the openssl command line application because it's faster and more reliable.  Instead of having to create/delete temporary files for the openssl command line program to read, we're storing the documents in memory and passing those directly to the verification routines.

There are a few levels of tests:

* C tests which verify that the `src/verify.c` and `src/verify.h` program are implemented correctly.
  * written in `src/tests.c`
  * run with and without electric fence and valgrind
  * do not leak
* Javascript tests which verify the higher level interface as well as the binding source `src/module.c`
  * There's a set of tests which go through and flip each bit independently of the document and the cryptographic portions of the pubkey and pkcs7 file, one bit at a time.  This is to verify that even a single bit flip anywhere in the document will result in a failing validation
* Bash tests which are just double checking that what is valid on the command line program is also valid in our program

These should run on linux, but it's possible they also run on Mac.  I don't know of any reasons off the top of my head why they wouldn't work on Windows, but compatibility there is not a concern.  Oh, and this is *way* faster than using the cli openssl, since there's no process creation or file i/o management overhead.

This is also my first N-API module and I'm not an expert in C, but I think I've done this right.

jhford commented:
Oh, and this is intended to be a general purpose library, which is why there's not a huge amount of tc-specific stuff going on in here.

jhford commented:
@franziskuskiefer hey, this is the module that we were talking about.  It'd be great if you had a chance to take a look and let me know if there's anything which scares you in it.  The `NOVERIFY` flag is required, per EC2, in order to operate with the signing that EC2 does.

djmitche commented on index.js:
```diff
@@ -0,0 +1,68 @@
+    if (pkcs7[i] !== pkcs7_header[i]) {
+      hasHeader = false;
+    }
+  }
```
`Buffer.compare` would probably be a bit more succinct and faster here.  Something like `pkcs7_header.compare(pkcs, 0, pkcs7_header.length) == 0`

jhford replied:
I'll change to using Buffer.compare.  The compare method on the instance was working in a strange way, but I doubt one is faster than the other, since .slice() should be a close-to-zero operation since it shares the same backing memory.

djmitche commented on index.js:
```diff
@@ -0,0 +1,68 @@
+ * which can be converted into a Buffer safely with a call to Buffer.from(),
+ * with an encoding parameter of 'utf-8'.
+ */
+module.exports = function verify(pubkey, document, pkcs7) {
```
In the README, the third argument is `rsa2048` -- which one is correct?

jhford replied:
In the readme, rsa2048 is a local variable to show the example usage, though, the function can take any PKCS#7 file.  In the code, pkcs7 is correct, as it's any PKCS#7.

djmitche replied:
I think the error message will be confusing for the user.  Can you find some way to make it clear that "pkcs7 is required" means the third argument?

djmitche commented on index.js:
```diff
@@ -0,0 +1,68 @@
+    }
+  }
+
+  // If the document has a header, we're going to assume that it's totally
```
"If the pkcs7 value has .."?

jhford replied:
yep :)

djmitche commented on binding.gyp:
```diff
@@ -1,8 +1,47 @@
+  },
+  'targets': [
+    {
+      'target_name': 'module',
```
This should probably have a more descriptive name?

djmitche commented on binding.gyp:
```diff
@@ -1,8 +1,47 @@
+    # node v0.6.x doesn't give us its build variables,
+    # but on Unix it was only possible to use the system OpenSSL library,
+    # so default the variable to "true", v0.8.x node and up will overwrite it.
+    'node_shared_openssl%': 'true'
```
Is this necessary, since we're targeting Node 10.x?

jhford replied:
Only the existence of the variable is not there in 0.6.x, but my understanding is that >0.8.x could be built using the system OpenSSL, and that variable will be set based on the build config.  This is also the documented approach for supporting using the same openssl as the Node runtime, per the node-gyp wiki

djmitche commented on src/module.c:
```diff
@@ -0,0 +1,101 @@
+  void *document;
+  void *signature;
+
+  status = napi_get_buffer_info(env, argv[0], &pubkey, (size_t *)&pubkey_l);
```
This function returns the size of the buffer, and overwrites the value returned from `npapi_get_value_int64` above.  Is there any need to pass the lengths as separate JS args?

If those `npapi_get_value_int64` calls are omitted, then the lengths can all be `size_t` and no casting would be required.  While it's likely that all of these fields are 64-bit, that wouldn't be the case on a 32-bit system or (in some mysterious future) a 128-bit system.  The world got itself in a lot of trouble assuming 32-bit pointers!  That would also address the uint64_t vs. int64_t issue above.

Passing lengths separately from the buffers also runs the risk of causing a segfault or reading unauthorized memory by passing a length larger than the buffer itself.

franziskuskiefer replied:
You really don't want to cast `uint64_t` to `size_t`.

jhford replied:
yep!  @djmitche i wrote this portion of the glue code thinking that I'd just mimic the C API into JS, then do this in JS.  Even though I did set the length in the napi_get_buffer_info call, I missed that I missed that this is the better approach.

I'll switch to using `size_t` and drop the extra parameters.

djmitche commented on src/verify.h:
```diff
@@ -0,0 +1,37 @@
+
+// Store the return value of the code.  Like command line tools, a value >=1
+// will be a failure and == 0 will be success.
+typedef unsigned int return_t;
```
perhaps `VF_result_t` just to avoid potential name conflicts?

jhford replied:
done

djmitche commented on src/verify.h:
```diff
@@ -0,0 +1,37 @@
+//
+// If there are errors encountered during the invocation, they will be stored
+// in the **errors list, along with the number of errors in the errors_n
+// parameter.
```
This should have some indication of memory ownership.  From what I can tell, the passed-in buffers remain the responsibility of the caller, and it is the caller's responsibility to free all of the Error objects in the returned linked list.

jhford replied:
So I'm allocating this memory on the stack instead of the heap, so it might actually be that the memory is being used after the stack has released it, and coincidentally working.  I will rethink this, but this might be more wrong than we both thought :)

djmitche replied:
Oh indeed!  Yikes!

djmitche commented on src/module.c:
```diff
@@ -0,0 +1,101 @@
+                     signature_l, &err);
+
+  // We should read all of these at some point, and maybe even
+  // give them all the information in the error nodes
```
I'm not sure what this means -- is it indicating that only the first error is displayed?

jhford replied:
I've rephrased this.

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+  OpenSSL_add_all_algorithms();
+  // Since none of these functions return useful error messages, per the
+  // openssl wiki documentation, we're just going to return success status.
+  // If this changes in future, we have the option to start failing
```
True, but the return value is ignored.  For consistency, it's probably best to return void from this function, or to handle a potential failure in init().

jhford replied:
done

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+//  - make sure all cleanup happens all the time
+//  - make sure that all return values are correct read
+//  - consider using ERR_get_error (iirc) to get actual error reasons
+//  - make sure we're correctly forcing FORMAT_PEM
```
Are these done?  If not, best to file issues for them rather than leave inline TODO comments..

jhford replied:
Yes, I'll delete the list

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+  // We don't want the BIO_close method here to free() the memory passed in
+  // BIO_set_close(bio_pubkey, BIO_NOCLOSE);
+  // BIO_set_close(bio_document, BIO_NOCLOSE);
+  // BIO_set_close(bio_pkcs7, BIO_NOCLOSE);
```
Why is this commented out?  I would expect that this would cause the BIO layer to try to free memory that belongs to JS..

jhford replied:
I was trying to diagnose a memory leak.  This code should be deleted, while I'll do.

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+  // option of the function works well
+  //
+  //   if(!PEM_read_bio_PKCS7(bio_pkcs7, &p7, NULL, NULL)) {
+  //
```
The docs say:

> The PEM read functions all take an argument TYPE **x and return a TYPE * pointer. Where TYPE is whatever structure the function uses. If x is NULL then the parameter is ignored. If x is not NULL but *x is NULL then the structure returned will be written to *x. If neither x nor *x is NULL then an attempt is made to reuse the structure at *x (but see BUGS and EXAMPLES sections). Irrespective of the value of x a pointer to the structure is always returned (or NULL if an error occurred).

I suspect that this crash is because `p7` is not initialized to `NULL`.  But using the return value is also fine, so you can probably just remove this comment :)

The lack of NULL initialization could also potentially cause a crash calling `PKCS7_free(..)` on a garbage value

franziskuskiefer replied:
Indeed. `p7` should be nulled on initialised independent on how you use it.

jhford replied:
ahh, that makes sense.  I'll `NULL` it.

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  // Same as reading the PKCS#7 file above
+  //   if (!PEM_read_bio_X509(bio_pubkey, &cert, 0, NULL)) {
```
Interesting that cert *is* initialized to NULL -- but perhaps that was added later?

jhford replied:
it was!

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  if (1 == PKCS7_verify(p7, certs, store, bio_document, NULL,
+                        PKCS7_NOINTERN | PKCS7_NOVERIFY)) {
```
The NOVERIFY here means to not verify the cert chain (right?).  A comment might clarify that for the reader..

jhford replied:
sure

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+      newError.next = head;
+      head = &newError;
+    } while (errorNum);
+  }
```
Rather than have `VF_verify` return a list of errors (only one of which is displayed), it would probably be better to use this function to concatenate together a string containing all of the errors in some vaguely human-readable fashion ("vaguely" because I've seen OpenSSL's errors before!).  Then the error output parameter from `VF_verify` would be a string with the responsibility of freeing that string on the caller.  As-is, it looks like the error list is leaked.

jhford replied:
Yes, it would be better, but I think that for the time being, this is enough.  Per the discussion above about how the memory for the Error list is allocated, I think that a bit of a rethink is needed here around how errors are handled.

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+end:
+  if (!BIO_free(bio_document) || !BIO_free(bio_pubkey) ||
+      !BIO_free(bio_pkcs7)) {
+    rv = VF_EXCEPTION;
```
Do these necessarily add an error to the ERR module?  As-written, if the result is `VF_EXCEPTION` but there are no errors, `module.c` will not throw anything, but will return false, suggesting to the JS caller that there were no errors but the document was invalid.

Unless OpenSSL guarantees there will be an error, it might make sense to include a case for VF_EXCEPTION with no errors, and return `"Unknown OpenSSL error occurred"` or something like that.

jhford replied:
I've done this.  Ideally, we would add to the OpenSSL error queue, but that's a great follow up for an interested person.

walac replied:
`||` is short-circuited, you might want to change to:
```c
int ret = BIO_free(bio_document);
ret &&= BIO_free(bio_pubkey);
ret &&= BIO_free(bio_pkcs7);
if (!ret) {
  rv = VF_EXCEPTION;
  ...
```

djmitche commented on test-cmdline.sh:
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
It looks like this just exercises OpenSSL, not the code in this module?  I'm confused what the purpose is..

walac replied:
Same here...

jhford replied:
the purpose is to show that files which validate using the OpenSSL binary also validate in this library.  If the library test fails, we know whether it's because the test-files are valid or not based on whether this test works or not.

djmitche replied:
It might be good to add some comments to that effect since at least two of us were confused.  I was particularly confused since this seems to benchmark openssl and run it under valgrind.  I can see the use of benchmarking in development, but since none of the code in this repo is involved, I don't understand the use of valgrind.

Perhaps make benchmarking and valgrind optional, so that running the tests doesn't wait for it?

djmitche commented on test.js:
```diff
@@ -0,0 +1,191 @@
+            parseInt(badDoc[i], 10).toString(2),
+            ' -> ',
+            parseInt(badDoc[i] ^ (1 << j), 10).toString(2),
+          ].join(' '))*/
```
leftover debugging code

jhford replied:
thanks, removed

djmitche commented on test.js:
```diff
@@ -0,0 +1,191 @@
+    // We're going to flip some bits here.  We're only going to try to flip the
+    // bits that relate to the cryptographic calculations.  We don't flip all
+    // bits for this document since we're not doing signature file validation
+    // because of a self-signed certificate being used (i.g. -noverify)
```
this is really cool!

jhford replied:
thanks :)  It was a fun set of tests to write

djmitche commented on test.js:
```diff
@@ -0,0 +1,191 @@
+    // We're going to flip some bits here.  We're only going to try to flip the
+    // bits that relate to the cryptographic calculations.  We don't flip all
+    // bits for this document since we're not doing signature file validation
+    // because of a self-signed certificate being used (i.g. -noverify)
```
repeated comment..

jhford replied:
updated to make them accurate

djmitche commented on src/tests.c:
```diff
@@ -0,0 +1,224 @@
+  return x;
+}
+
+// Do some tests.  I don't care about leaking memory in this function
```
It can be a little more welcoming to use "we" in comments instead of 'I", or just avoid a pronoun altogether -- otherwise a reader looks at this code and thinks, "Oh, this belongs to someone, I shouldn't touch it".  So perhaps "Memory leaks in this function are not important, since this is just a test harness, but covering the tested code is very important".

jhford replied:
Most removed, left in testing code where it's difficult to reword

djmitche commented on README.md:
```diff
@@ -0,0 +1,53 @@
+message is signed with a self-signed certificate.
+
+# Developing
+You'll need Linux.  Other systems might work, but are completely untested.
```
You may want to list additional requirements here.  I didn't have:
 * clang
 * clang-format
 * valgrind
 * electric-fence

franziskuskiefer commented on src/module.c:
```diff
@@ -0,0 +1,101 @@
+
+  void *pubkey;
+  void *document;
+  void *signature;
```
Why are these all void pointer? They should have a real type.

franziskuskiefer commented on src/module.c:
```diff
@@ -0,0 +1,101 @@
+
+  status = napi_get_buffer_info(env, argv[0], &pubkey, (size_t *)&pubkey_l);
+  if (status != napi_ok) {
+    napi_throw_error(env, NULL,
```
You don't free anything in any of these returns. I suspect you're leaking some of your pointers in error cases.

franziskuskiefer commented on src/module.c:
```diff
@@ -0,0 +1,101 @@
+  uint64_t pubkey_l;
+  uint64_t document_l;
+  uint64_t signature_l;
+  status = napi_get_value_int64(env, argv[1], (int64_t *)&pubkey_l);
```
This doesn't seem like a wise cast.

walac replied:
nit: Why not declare `pubkey_l` as an `int64_t`?

franziskuskiefer commented on src/tests.c:
```diff
@@ -0,0 +1,224 @@
+  }
+
+  long size = ftell(f);
+  *length = size;
```
You shouldn't assign any output variables before you know that everything passed. This will return a value for `length` even if an error occurred.

franziskuskiefer commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+//  - consider using ERR_get_error (iirc) to get actual error reasons
+//  - make sure we're correctly forcing FORMAT_PEM
+return_t VF_verify(char *pubkey, uint64_t pubkey_l, char *document,
+                   uint64_t document_l, char *pkcs7, uint64_t pkcs7_l,
```
I'd always recommend to use `unint8_t` for bytes such as the `pkcs7` or `document`.

franziskuskiefer commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+//  - make sure that all return values are correct read
+//  - consider using ERR_get_error (iirc) to get actual error reasons
+//  - make sure we're correctly forcing FORMAT_PEM
+return_t VF_verify(char *pubkey, uint64_t pubkey_l, char *document,
```
Lengths should usually be `size_t`. You don't want to process input that has a length needing 64-bit.

franziskuskiefer commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+  // We want to clear the OpenSSL Error queue so that we know when we're in the
+  // cleanup section, any errors we hit are the result of this invocation
+  ERR_clear_error();
+  return_t rv;
```
`rv` isn't used until way later. Initialise closer to its use.

walac commented on Makefile:
```diff
@@ -0,0 +1,22 @@
+	clang-format -i src/*.c src/*.h
+
+
+.PHONY: test
```
I am not sure if `.PHONY` is an addictive target, the code I have seen put everything into one target definition, like so:
```make
.PHONY: memtests ctests shell-tests format test
```

jhford replied:
This is a stylistic preference.  I put them together because if I remove the phony-target, the diff is localised to the area of the file where the rule is.

djmitche replied:
Fair enough.  It works in my testing, too.

walac commented on Makefile:
```diff
@@ -0,0 +1,22 @@
+.PHONY: memtests
+memtests: src/verify.c src/tests.c
+	clang -g $? -o ./$@ -Wall -Werror -lcrypto -lefence -DEXTRA_DEBUG -DBENCH_ITER=$(BENCH_ITER)
+	valgrind --leak-check=full ./$@
```
I typically use some useful extra options: `--read-var-info=yes --track-origins=yes`

jhford replied:
added, but no new errors found :)

walac commented on Makefile:
```diff
@@ -0,0 +1,22 @@
+	./$@
+
+.PHONY: shell-tests
+	./test-cmdline.sh
```
Did you mean
```make
.PHONY: shell-tests
shell-tests:
        ./test-cmdline.sh
```
?

jhford replied:
yes

walac commented on index.js:
```diff
@@ -0,0 +1,68 @@
+  if (!hasHeader) {
+    while (pkcs7.length > 0 && pkcs7.indexOf(nl) === 0) {
+      pkcs7 = pkcs7.slice(1);
+    }
```
I think this will make js VM happier:

```javascript
let i;
for (i = 0; i < pkcs7.length && pkcs7[i] === nl; ++i);
pkcs7 = pkcs7.slice(i);
```

jhford replied:
This works, good idea.

walac commented on index.js:
```diff
@@ -0,0 +1,68 @@
+
+    while (pkcs7.length > 0 && pkcs7.lastIndexOf(nl) === pkcs7.length - 1) {
+      pkcs7 = pkcs7.slice(0, pkcs7.length - 1);
+    }
```
Same as above:

```javascript
for (i = pkcs7.length - 1; i >= 0 && pkcs7[i] === nl; --i);
pkcs7 = pkcs7.slice(i + 1, pkcs7.length);
```

jhford replied:
This one didn't work, but i modified it to
```javascript
53     for (i = pkcs7.length - 1; i >= 0 && pkcs7[i] === nl[0]; i--);
54     pkcs7 = pkcs7.slice(0, i+1); 
```

walac commented on src/module.c:
```diff
@@ -0,0 +1,101 @@
+  status = napi_get_cb_info(env, info, &argc, argv, NULL, NULL);
+
+  if (status != napi_ok) {
+    napi_throw_error(env, NULL, "Failed to parse arguments");
```
Doesn't it need a return statement here? I believe `napi_throw_error` raises an exception in the js environment, not in the C code.

jhford replied:
I couldn't find any examples where there was a return after calling a napi_throw* function, but you're right that it does continue C execution.  I've either removed all of my napi_throw* calls, added a return NULL/status where appropriate or left them without a return statement because there's nothing else happening in the function of import.

walac commented on src/module.c:
```diff
@@ -0,0 +1,101 @@
+  // give them all the information in the error nodes
+  if (err != NULL) {
+    char errMsg[256];
+    if (!snprintf(errMsg, 256, "%s@%d#%s: %s", err->file_string, err->line,
```
In case of error, `snprintf` returns a negative number.

walac commented on src/tests.c:
```diff
@@ -0,0 +1,224 @@
+
+  int read = fread(*value, 1, size, f);
+
+  if (!ferror(f) || read != size) {
```
I think the correct logic is: `!ferror(f) && read == size`

walac commented on src/tests.c:
```diff
@@ -0,0 +1,224 @@
+
+  if (!ferror(f) || read != size) {
+    fclose(f);
+    return VF_SUCCESS;
```
Potential buffer overrun here: `malloc` doesn't zero the allocated memory, so you need either:
* use [calloc](https://en.cppreference.com/w/c/memory/calloc) instead or
* set `*value[size] = '\0'`

jhford replied:
it needs `(*value)[size] = 0;`, with the bracket around the dereference :)

walac commented on src/tests.c:
```diff
@@ -0,0 +1,224 @@
+  free(value);
+}
+
+char *memdup(char *src, long len) {
```
nit: use `const char *`

walac commented on src/tests.c:
```diff
@@ -0,0 +1,224 @@
+
+// Returns the number of bytes read, and sets the contents ** to the start
+// of the memory buffer
+return_t read_complete_file(char *filename, char **value, long *length) {
```
nit: use `size_t` for length values

walac commented on src/verify.c:
```diff
@@ -0,0 +1,131 @@
+      newError.func_string = ERR_func_error_string(errorNum);
+      newError.next = head;
+      head = &newError;
+    } while (errorNum);
```
Since we will never reach the end of the loop with `errorNum` falsy, we could replace the `do while` for a `for(;;)`.

walac commented on src/verify.h:
```diff
@@ -0,0 +1,37 @@
+#ifndef VERIFY_H
+#define VERIFY_H
```
nit: this can be replaced by the more elegant `#pragma once`

walac commented on src/verify.h:
```diff
@@ -0,0 +1,37 @@
+typedef unsigned int return_t;
+#define VF_SUCCESS 0
+#define VF_FAIL 1
+#define VF_EXCEPTION 2
```
It is a POSIX convention to return error results as negative numbers.

djmitche commented:
It's worth saying again, there are lots of comments on this PR but none of them are huge changes.  This really is nicely done and I'm looking forward to seeing it released.  Maybe you can even convince AWS to link to it from their docs?

jhford commented:
I've addressed the feedback appropriately.  I've essentially rewritten the entire Error handling section of the library because there were some pretty serious problems with how that was being done.

I'm dismissing review requests so that we can do another round of review, too bad github uses such rough language around that ;)

djmitche commented:
btw if you just re-select a reviewer for review it will avoid that language (just "requested review from..")

djmitche commented on README.md:
```diff
@@ -0,0 +1,62 @@
+// http://169.254.169.254/latest/dynamic/instance-identity/{rsa2048, document}
+// Public key provided by Amazon
+let document = fs.readFileSync('document');
+let rsa2048 = fs.readFileSync('rsa2048');
```
Is this a signature for the document?  Perhaps calling the local variable `signature` and also referring to `signature` in the error message?

jhford replied:
The readme is targetting users of the library, which are suggested to download the endpoint `rsa2048`.  It's not a named argument, so it doesn't matter if the variable in the readme example code matches the code in index.js.

Example code here is the suggested use case, which is the rsa2048 endpoint, but the code could work with other files.

djmitche replied:
It's the error message I'm worried about: "pkcs7 must be provided' -- what is pkcs7?

jhford replied:
I'll add a lot of clarification in the README file as to what each of these things are.

djmitche commented on src/glue.c:
```diff
@@ -0,0 +1,203 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+//#define VF_DEBUG 1
```
An option here might be
```c
#if 1
#define VF_DEBUG(...)
#else
#define VF_DEBUG(...) printf(__VA_ARGS__)
#endif
```

This would avoid the need for lots of #ifdef's below.  The #ifdef's make the code surprisingly hard to read since they interrupt the flow of the indentation.

jhford replied:
I played around with that, but I was having trouble figuring out what to do for the empty macro case... CPP is a funny program sometimes :)

walac replied:
Try this:
```c
#define VF_DEBUG(...) do {} while(0)
```

jhford replied:
it works without having a value defined for the macros with both gcc 8.1.1 and clang 6.0.0, and it's a little cleaner than an empty do/while I think :)

walac replied:
The goal of empty do while is for cases like so:
```cpp
if (condition)
   VF_DEBUG("text");
```

jhford replied:
ahh, cool.  i guess that's something the compiler will optimize out pretty easily

djmitche commented on src/glue.c:
```diff
@@ -0,0 +1,203 @@
+  printf("created js Error\n");
+#endif
+
+  // We're limiting to the first 1000 so that we don't
```
1000 (here) or 10 (below)?  It would probably be good to indicate that errors were dropped when hitting the limit (for example, adding "..subsequent errors omitted").

Also, `VF_err_free` will hang forever if this is a circular linked-list, so I'm not sure this check helps..

jhford replied:
I've replaced the numbers with a macro `MAX_OPENSSL_ERRORS` for the display and set it to 25.  The difference is that `VF_err_free` is required in order to avoid a memory leak.  It's also something defined in the same file, and building the list is tested in the unit tests.

jhford replied:
I've added a specific unit test to ensure that the list is built and consumed correctly

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,182 @@
+                  err->reason);
+
+  if (size < 0) {
+    return "ERROR DETERMINING LENGTH FOR ERROR STRING";
```
why is this message (and the one below) capitalized?

jhford replied:
yah, you're right that it isn't the right thing to do.  I had forgotten about the pattern of returning a pointer to NULL on error in C.  I'll switch it to return NULL on error and then handle that from the HandleError function.

(comment edited to better convey my original intention)

djmitche commented on src/verify.h:
```diff
@@ -0,0 +1,53 @@
+//
+// If there are errors encountered during the invocation, they will be stored
+// in the **errors list, along with the number of errors in the errors_n
+// parameter.  This memory is allocated in the VF_verify function but is owned
```
There appears to be no `errors_n` parameter.

jhford replied:
I have removed that parameter from the comment since we do not set it.

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,182 @@
+      // shouldn't have been any removals from the error queue.  Sadly, this is
+      // not the case, since the call to ERR_get_error_line here returns a
+      // falsy value, which means that the error queue doesn't work as
+      // documented in the man page.
```
Maybe I'm misunderstanding this comment, but the behavior makes sense to me.  Assuming there was only one error in the queue, on the first iteration of the `do-while`, this call will pop the entry from the queue.  On the second iteration, it will return 0.  Continuing on to call `ERR_reason_error_string` for that 0 would be an error, so you need to exit the loop here (and the `while (errNum)` at the end is redundant.

jhford replied:
I'll modify it to a `for(;;)` loop.  I think I'd been looking at the code too long when I wrote that.  I thought it was an error in the interaction between the previous peek() call and this get() call, but you're right here.

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,182 @@
+    head->lib = __FILE__;
+    head->func = "VF_verify";
+    head->next = NULL;
+    head->line = __LINE__;
```
`head->file` is not set here, leaving it at an undefined value

djmitche replied:
and correspondingly, `VF_err_fmt` omits the filename, so this never crashes anything (until someone notices and fixes that..)

jhford replied:
yep!  I've added that and now including that in the error messages.  That said, snprintf versions that I have been using do not crash when a `char*` is `NULL`:
```C
[jhford@jhford-desktop iid-verify]$ cat printf.c 
#include <stdio.h>
#include <stdlib.h>

int main(void) {
  struct E {
    char* d;
  };
  struct E *e = malloc(sizeof(struct E));
  e->d = NULL;

  char *msg = malloc(1001);
  snprintf(msg, 1000, "%s\n", e->d);
  printf("%s", msg);
  return 0;
}
[jhford@jhford-desktop iid-verify]$ clang -Wall -Werror printf.c -o test
[jhford@jhford-desktop iid-verify]$ ./test 
(null)
[jhford@jhford-desktop iid-verify]$ echo $?
0
[jhford@jhford-desktop iid-verify]$
```

djmitche replied:
The code uses `malloc` to allocate the memory for the Error, so the `file` field contained whatever 64 bits were in that location already -- not necessarily NULL.  But now it's initialized so we're all good :)

jhford replied:
heh, yah, you're right.  I always forget that C does not initialize to 0s on malloc.

djmitche commented on src/verify.c:
```diff
@@ -0,0 +1,182 @@
+    // verification error or not.  If it is, mark this invocation of VF_verify
+    // as VF_FAIL, then clear the error queue so that if future errors occur,
+    // handle them as exceptions.  This also ensures that exceptions during
+    // verification are handled differently to invalid signatures
```
Good thinking!

jhford replied:
thanks :)

djmitche commented on README.md:
```diff
@@ -11,8 +11,21 @@ let verify = require('iid-verify');
+
+// document is a JSON file available through the metadata service // which
+contains information about the instance like instance id and // AMI used to
+configure the instance
```
comments got wrapped here

djmitche commented on README.md:
````diff
@@ -24,17 +37,62 @@ if (verify(pubkey, document, rsa2048)){
+```javascript
+let err = new Error('asn1 encoding routines ../deps/openssl/openssl/crypto/asn1/asn1_lib.c:157 ASN1_get_object header too long');
+err.errors = [
+'PEM routines ../deps/openssl/openssl/crypto/pem/pem_oth.c:83 PEM_ASN1_read_bio ASN1 lib',
````
indentation

djmitche commented on index.js:
```diff
@@ -20,7 +20,7 @@ module.exports = function verify(pubkey, document, pkcs7) {
 
   if (typeof pkcs7 === 'undefined') {
-    throw new Error('pkcs7 must be provided');
+    throw new Error('pkcs7 signature must be provided');
```
++

jhford commented:
@djmitche Cool!  You're right about the VF_err_free call in the HandleError method possibly hanging.  I was thinking about this a little more, and I think that the most likely case for this to be infinite is that someone is incorrectly reusing memory, either a stack struct or heap memory allocated outside of the loop that builds the linked list.  I hit both in my development.  It occurred to me just now that while traversing the linked list, the most simple check would be that when it's time to assign the list head to the next value, check that they are different pointers.  That would remove any need for the 25 error constant and would also help in VF_err_free.

I've added this check and removed the constant, please let me know if that changes the r+

djmitche commented:
It doesn't change the r+.  In general defensive programming is good, but there's not much to be done if a linked list is incorrectly linked, since it's such a simple data structure.  So I'd be inclined to just assume it's correct at runtime and that any issues will be caught in tests via a crash or hang.  But extra checking doens't hurt.

jhford commented:
@walac any thoughts?

walac commented on binding.gyp:
```diff
@@ -1,8 +1,48 @@
+      'target_name': 'glue',
+      'cflags': [
+        '-std=c99',
+        '-Wall',
```
If you want to be really picky, can add `-Wextra` too

jhford replied:
Done, made a slight change because GCC was unhappy with a switch's case falling through to default.

walac commented on src/glue.c:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  int i = 0;
+  while (err != NULL) {
```
I think a `for` is better suited here:
```c
for (int i = 0; err; ++i) {
```

jhford replied:
sure, changed to this structure.

walac commented on src/glue.c:
```diff
@@ -0,0 +1,163 @@
+    status = napi_create_string_utf8(env, msg, NAPI_AUTO_LENGTH, &errorString);
+    if (status != napi_ok) {
+      VF_ERROR("could not create js string for openssl error: %s\n", msg);
+      return status;
```
memory leak here, we need to `free(msg)`

jhford replied:
fixed

walac commented on src/glue.c:
```diff
@@ -0,0 +1,163 @@
+
+    if (msg == NULL) {
+      VF_ERROR("could not format Error struct %s\n", err->reason);
+      return napi_generic_failure;
```
memory leak here, we need to `free(msg)`

jhford replied:
fixed

walac commented on src/glue.c:
```diff
@@ -0,0 +1,163 @@
+    if (err == err->next) {
+      VF_ERROR("found simple-cycle in error linked list\n");
+      break;
+    }
```
Is there a special reason for this check?

jhford replied:
yes.  as dicussed earlier, this is a simple cycle checker.  It detects the easiest way to have a badly constructed linked list, for example one where each node was constructed using the same heap memory.  Since this is a simple check, it's worth having.

walac commented on src/glue.c:
```diff
@@ -0,0 +1,163 @@
+  uint8_t *document;
+  uint8_t *signature;
+
+  status = napi_get_buffer_info(env, argv[0], (void *)&pubkey, &pubkey_l);
```
The correct cast is `(void **)`, isn't it? Or it could be `&(void *)pubkey`
But maybe `pubkey`, `document` and `signature` should all be `void *`?

jhford replied:
I'm not sure either, but both `(void *)&` and `(void **)&` work equally well.  `&(void*)` is a compiler error though.  Since this is getting information from a Buffer (Uint8Array) from JS, it's probably best to match that in C.

walac commented on src/glue.c:
```diff
@@ -0,0 +1,163 @@
+  if (result == VF_EXCEPTION) {
+    status = HandleError(env, err);
+    if (status != napi_ok) {
+      napi_throw_error(env, NULL, "could not handle error");
```
If it fails, `HandleError` doesn't free `err`

jhford replied:
Fixed

walac commented on src/tests.c:
```diff
@@ -0,0 +1,287 @@
+
+  struct Error *err = NULL;
+
+  gettimeofday(&start, NULL);
```
`gettimeofday` isn't the best time function for performance measurement because it doesn't account for thread switches, [clock_gettime](https://linux.die.net/man/3/clock_gettime) with `CLOCK_THREAD_CPUTIME_ID` clock id is a better option.

jhford replied:
Cool!  thanks for the pointer to that function.  For the first version, gettimeofday is probably good enough, especially since we're not doing any thread switching in these unit tests.

walac replied:
Notice I am talking about thread system in a system level. Say, in the middle of the verify function, the kernel switches to a kernel thread responsible to flush system memory pages, `gettimeofday` will account for that time, while `clock_gettime` won't.

g-k commented on README.md:
```diff
@@ -0,0 +1,120 @@
+  * clang-format
+  * valgrind
+  * electric-fence
+
```
Might be worth enumerating known working commands for different distros and versions like "on ubuntu 18.04 run: `sudo apt-get install clang clang-format electric-fence valgrind`"

jhford replied:
Good idea, I'll also do Fedora.

g-k commented on index.js:
```diff
@@ -0,0 +1,68 @@
+    pkcs7 = Buffer.from(pkcs7, 'utf-8');
+  }
+
+  hasHeader = false;
```
nit: add a `let =` here?

jhford replied:
good catch.  I've actually added a 'use strict' directive now, thanks for the reminder :)

g-k commented on test.js:
```diff
@@ -0,0 +1,227 @@
+            // and invalid PEM file
+            let asn1Errors = [
+              /ASN1_CHECK_TLEN/,
+              /ASN1_ITEM_EX_D2I/
```
nit: eslint was adding a trailing comma for me here.

jhford replied:
That's the taskcluster coding style, but I'll commit that fix

g-k commented on test.js:
```diff
@@ -0,0 +1,227 @@
+  });
+
+  describe('with bad values', () => {
+    it('should return false with empty Buffer values', () => {
```
nit: looks like the test descriptions got out of sync and we can replace "return false" with "throw error" in this describe block

jhford replied:
you're right, thanks!

g-k commented on index.js:
```diff
@@ -0,0 +1,68 @@
+ * signature.  All parameters should either be strings, Buffers or something
+ * which can be converted into a Buffer safely with a call to Buffer.from(),
+ * with an encoding parameter of 'utf-8'.
+ */
```
nit: comment doesn't mention what the function returns or errors it can throw; consider linking to the #API or #Errors section of the readme

jhford replied:
Good idea.  I've included a text reference, but when its landed, I'll insert links.

g-k commented on README.md:
```diff
@@ -0,0 +1,120 @@
+2. Document does not validate against signature -- instance identity document,
+   public key or pkcs7 signature is not correct.  `verify` returns false.
+3. A javascript `Error` is thrown from the library along with information about
+   which error occurred.
```
nit: seems complicated to return `isValid` and throw errors. Is returning an array of errors or throwing an "invalid signature" error an option?

jhford replied:
It's been a taskcluster pattern to have checks like these use a boolean true/false to represent the two successful validation cases and thrown errors only when the validation itself could not be completed.  I tend to agree with you, but I wanted to match conventions elsewhere in our code.

g-k commented on Makefile:
```diff
@@ -0,0 +1,29 @@
+.PHONY: memtests
+memtests: src/verify.c src/tests.c
+	clang -g $? -o ./$@ -Wall -Wextra -Werror -lcrypto -lefence -DEXTRA_DEBUG -DBENCH_ITER=$(BENCH_ITER)
+	valgrind --read-var-info=yes --track-origins=yes --leak-check=full ./$@
```
nit: might be nice to have some additional sanitizers and fuzzing too:

* https://clang.llvm.org/docs/AddressSanitizer.html
* https://clang.llvm.org/docs/MemorySanitizer.html
* https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
* fuzzing-project or afl

jhford replied:
neat!  I'll look into those and see if they work with the code or at least get an idea of what's required.

jhford commented:
@g-k no worries at all, I am happy to have as many eyes on this code as I can get!
